From 2ea812a006ae2310fb17a6ac3d5ff67498ddcd1c Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Thu, 26 Nov 2020 03:57:23 +0100
Subject: [PATCH 1/1] fs: fat: ensure flushing of directory

If we allocate a new cluster for the directory, we must ensure that it is
flushed to disk.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 fs/fat/fat_write.c | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/fs/fat/fat_write.c b/fs/fat/fat_write.c
index a3c3085206..4b407fbef1 100644
--- a/fs/fat/fat_write.c
+++ b/fs/fat/fat_write.c
@@ -1167,10 +1167,20 @@ static dir_entry *find_directory_entry(fat_itr *itr, char *filename)
 
 	/* allocate a cluster for more entries */
 	if (!itr->dent &&
-	    (!itr->is_root || itr->fsdata->fatsize == 32) &&
-	    new_dir_table(itr))
-		/* indicate that allocating dent failed */
-		itr->dent = NULL;
+	    (itr->is_root || itr->fsdata->fatsize == 32)) {
+
+		if (new_dir_table(itr)) {
+			/* indicate that allocating dent failed */
+			itr->dent = NULL;
+			return NULL;
+		}
+
+		/* Write directory table to device */
+		if (flush_dir(itr)) {
+			itr->dent = NULL;
+			return NULL;
+		}
+	}
 
 	return NULL;
 }
-- 
2.29.2

